keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,1169 @@
+
+
+> These tests will very closely resemble what we would see in production 
+> environment since these tests exercise the entire handler chain of the
+> kube-apiserver
+
+> We can also use the tests where `PerHandlerReadWriteTimeout == false` to document
+> what the client sees today with the timeout filter enabled, and compare with the 
+> results from tests where `PerHandlerReadWriteTimeout == true` to understand how 
+> differently the client is impacted with this new feature enabled.
+
+
+###### Request Handler Should Terminate
+
+
+Additionally, we should have test(s) that verify that a blocked handler does
+not block the client, with both the feature enabled/disabled and 
+for http/1x and http/2.0.
+```
+neverDoneCh := make(chan struct{})
+func(w http.ResponseWriter, req *http.Request) {
+	<-neverDoneCh
+}
+```
+
+###### Fake ResponseController
+
+
+Use a fake `ResponseController` to assert on whether the desired read/write
+timeout duration is provided, see: https://github.com/golang/go/issues/60229
+```
+// WithFakeResponseController extends a given httptest.ResponseRecorder object
+// with a fake implementation of http.ResonseController.
+func WithFakeResponseController(w *httptest.ResponseRecorder) *FakeResponseRecorder {
+	return &FakeResponseRecorder{ResponseRecorder: w}
+}
+
+type FakeResponseRecorder struct {
+	*httptest.ResponseRecorder
+
+	ReadDeadlines  []time.Time
+	WriteDeadlines []time.Time
+}
+
+func (w *FakeResponseRecorder) SetReadDeadline(deadline time.Time) error {
+	w.ReadDeadlines = append(w.ReadDeadlines, deadline)
+	return nil
+}
+
+func (w *FakeResponseRecorder) SetWriteDeadline(deadline time.Time) error {
+	w.WriteDeadlines = append(w.WriteDeadlines, deadline)
+	return nil
+}
+```
+
+##### Integration tests
+
+- Scenario A: `PerHandlerReadWriteTimeout` is disabled, a request times out
+- Scenario A: `PerHandlerReadWriteTimeout` is enabled, a request times out
+
+We exercise the following scenario with both http/1x and http/2.0 using client-go.
+
+
+
+##### e2e tests
+
+Similar to Integration tests
+
+### Graduation Criteria
+
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- All new tests enumerated in "Test Plan" are implemented
+
+#### Beta
+
+TBD
+
+#### GA
+
+TBD
+
+#### Deprecation
+
+
+### Upgrade / Downgrade Strategy
+
+Not applicable
+
+### Version Skew Strategy
+
+Not applicable
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [ ] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: PerHandlerReadWriteTimeout
+  - Components depending on the feature gate:
+	  - kube-apiserver
+
+###### Does enabling the feature change any default behavior?
+
+An API client will no longer receive the `504 GatewayTimeout` HTTP status code
+when the server times out a request. Instead, the client will receive an `error` 
```

Comment:
Is handling these errors more expensive than `504` timeouts? At least for HTTP1, I guess the client will typically `connect()` again (and perform another TLS handshake, etc.)?

I'm curious about pathological situations that could be systemically worsened if instead of receiving a `504` and retrying on an existing socket, clients instead need to reconnect.

Reply:
for HTTP/1x, the connection will be closed, so the client needs to reconnect. For http/2.0, the stream will be reset, and the tcp connection will still be intact.
I will add a test to document this behavior.